File: nvmolkit/tfd.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""GPU-accelerated Torsion Fingerprint Deviation (TFD) calculation.

This module provides GPU-accelerated implementations of TFD calculation,
compatible with RDKit's TorsionFingerprints API.

TFD (Torsion Fingerprint Deviation) is a measure of conformational similarity
based on the comparison of torsion angles between conformers. It was described
in: Schulz-Gasch et al., JCIM, 52, 1499-1512 (2012).

Example usage:
    >>> from rdkit import Chem
    >>> from rdkit.Chem import AllChem
    >>> import nvmolkit.tfd as tfd
    >>>
    >>> mol = Chem.MolFromSmiles('CCCCC')
    >>> AllChem.EmbedMultipleConfs(mol, numConfs=5)
    >>>
    >>> # Python lists (RDKit-compatible, default)
    >>> tfd_matrix = tfd.GetTFDMatrix(mol)
    >>>
    >>> # Numpy arrays
    >>> tfd_matrices = tfd.GetTFDMatrices(mols, return_type="numpy")
    >>>
    >>> # GPU tensors (fastest, no D2H copy)
    >>> tfd_matrices = tfd.GetTFDMatrices(mols, return_type="tensor")
"""

from contextlib import nullcontext

import numpy as np
import torch

try:
    import nvtx as _nvtx

    def _nvtx_range(name, color="blue"):
        return _nvtx.annotate(name, color=color)
except ImportError:

    def _nvtx_range(name, color="blue"):
        return nullcontext()


from nvmolkit import _TFD
from nvmolkit._arrayHelpers import *  # noqa: F403
from nvmolkit.types import AsyncGpuResult


class _TFDGpuResult:
    """Internal result container for GPU-resident TFD computation."""

    def __init__(self, tfd_values: AsyncGpuResult, output_starts: list[int]):
        self.tfd_values = tfd_values
        self.output_starts = output_starts

    def to_tensors(self) -> list[torch.Tensor]:
        """Extract as list of GPU tensors (no D2H copy).

        The TFD kernel runs on the generator's private CUDA stream. Returned
        tensors are valid for GPU-side ops ordered after this call on the same
        stream. For CPU-side access (e.g. ``tensor.cpu()``, ``tensor.item()``,
        ``tensor.numpy()``), call ``torch.cuda.synchronize()`` first to ensure
        the kernel has completed.
        """
        n = len(self.output_starts) - 1
        all_values = self.tfd_values.torch()
        return [all_values[self.output_starts[i] : self.output_starts[i + 1]] for i in range(n)]

    def to_numpy(self) -> list[np.ndarray]:
        """Extract as list of numpy arrays (one bulk D2H copy)."""
        n = len(self.output_starts) - 1
        torch.cuda.synchronize()
        all_values = self.tfd_values.numpy()
        return [all_values[self.output_starts[i] : self.output_starts[i + 1]] for i in range(n)]

    def to_lists(self) -> list[list[float]]:
        """Extract as Python lists (bulk D2H + tolist)."""
        n = len(self.output_starts) - 1
        torch.cuda.synchronize()
        all_list = self.tfd_values.numpy().tolist()
        return [all_list[self.output_starts[i] : self.output_starts[i + 1]] for i in range(n)]


def _get_gpu_result(mols, useWeights, maxDev, symmRadius, ignoreColinearBonds):
    """Run GPU TFD computation and return _TFDGpuResult (no D2H copy)."""
    pyarray, output_starts = _TFD.GetTFDMatricesGpuBuffer(
        mols,
        useWeights=useWeights,
        maxDev=maxDev,
        symmRadius=symmRadius,
        ignoreColinearBonds=ignoreColinearBonds,
    )
    return _TFDGpuResult(
        tfd_values=AsyncGpuResult(pyarray),
        output_starts=output_starts,
    )


def _extract_gpu_result(gpu_result, return_type):
    """Extract results from _TFDGpuResult based on return_type."""
    if return_type == "tensor":
        return gpu_result.to_tensors()
    elif return_type == "numpy":
        return gpu_result.to_numpy()
    elif return_type == "list":
        return gpu_result.to_lists()
    else:
        raise ValueError(f"Invalid return_type {return_type!r}. Must be 'list', 'numpy', or 'tensor'.")


def GetTFDMatrices(
    mols: list,
    useWeights: bool = True,
    maxDev: str = "equal",
    symmRadius: int = 2,
    ignoreColinearBonds: bool = True,
    return_type: str = "list",
) -> list[list[float]] | list[np.ndarray] | list[torch.Tensor]:
    """Calculate TFD matrices for multiple molecules.

    Args:
        mols: list of RDKit molecules, each with multiple conformers.
        useWeights: If True (default), use distance-based torsion weights.
        maxDev: Normalization mode ('equal' or 'spec').
        symmRadius: Radius for atom invariants (default: 2).
        ignoreColinearBonds: If True (default), ignore colinear bonds.
        return_type: Output format:
            'list' (default): list of Python lists (RDKit-compatible).
            'numpy': list of numpy float32 arrays.
            'tensor': list of GPU torch.Tensors (no D2H copy).

    Returns:
        list of TFD matrices in the requested format.
    """
    gpu_result = _get_gpu_result(mols, useWeights, maxDev, symmRadius, ignoreColinearBonds)
    with _nvtx_range("GPU: split per-molecule results", color="yellow"):
        return _extract_gpu_result(gpu_result, return_type)


def GetTFDMatrix(
    mol,
    useWeights: bool = True,
    maxDev: str = "equal",
    symmRadius: int = 2,
    ignoreColinearBonds: bool = True,
    return_type: str = "list",
) -> list[float] | np.ndarray | torch.Tensor:
    """Calculate the TFD matrix for conformers of a molecule.

    Convenience wrapper over GetTFDMatrices for a single molecule.

    Args:
        mol: RDKit molecule with multiple conformers.
        useWeights: If True (default), use distance-based torsion weights.
        maxDev: Normalization mode ('equal' or 'spec').
        symmRadius: Radius for atom invariants (default: 2).
        ignoreColinearBonds: If True (default), ignore colinear bonds.
        return_type: Output format:
            'list' (default): Python list of floats (RDKit-compatible).
            'numpy': numpy float32 array.
            'tensor': GPU torch.Tensor (no D2H copy).

    Returns:
        Lower triangular TFD matrix as a flat list, numpy array, or GPU tensor.
    """
    results = GetTFDMatrices(
        [mol],
        useWeights=useWeights,
        maxDev=maxDev,
        symmRadius=symmRadius,
        ignoreColinearBonds=ignoreColinearBonds,
        return_type=return_type,
    )
    if not results:
        if return_type == "numpy":
            return np.array([], dtype=np.float32)
        elif return_type == "tensor":
            return torch.tensor([], dtype=torch.float32)
        return []
    return results[0]

File: nvmolkit/nanobind/tfd.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

#include <GraphMol/ROMol.h>
#include <nanobind/nanobind.h>
#include <nanobind/stl/string.h>
#include <nanobind/stl/unique_ptr.h>

#include <stdexcept>
#include <string>
#include <vector>

#include "nvmolkit/nanobind/array_helpers.h"
#include "src/tfd/tfd_gpu.h"
#include "src/utils/nvtx.h"

namespace {

namespace nb = nanobind;
using namespace nb::literals;

std::vector<const RDKit::ROMol*> extractMolecules(const nb::list& molecules) {
  std::vector<const RDKit::ROMol*> converted;
  converted.reserve(molecules.size());
  for (std::size_t index = 0; index < molecules.size(); ++index) {
    const RDKit::ROMol* molecule = nb::cast<const RDKit::ROMol*>(molecules[index]);
    if (molecule == nullptr) {
      throw std::invalid_argument("Invalid molecule at index " + std::to_string(index));
    }
    converted.push_back(molecule);
  }
  return converted;
}

nvMolKit::TFDComputeOptions buildOptions(bool               useWeights,
                                         const std::string& maxDeviation,
                                         int                symmetryRadius,
                                         bool               ignoreColinearBonds) {
  nvMolKit::TFDMaxDevMode maxDeviationMode;
  if (maxDeviation == "equal") {
    maxDeviationMode = nvMolKit::TFDMaxDevMode::Equal;
  } else if (maxDeviation == "spec") {
    maxDeviationMode = nvMolKit::TFDMaxDevMode::Spec;
  } else {
    throw std::invalid_argument("maxDev must be 'equal' or 'spec', got: " + maxDeviation);
  }

  return {
    .useWeights          = useWeights,
    .maxDevMode          = maxDeviationMode,
    .symmRadius          = symmetryRadius,
    .ignoreColinearBonds = ignoreColinearBonds,
  };
}

nvMolKit::TFDGpuGenerator& gpuGenerator() {
  static nvMolKit::TFDGpuGenerator generator;
  return generator;
}

nb::tuple getTFDMatricesGpuBuffer(const nb::list&    molecules,
                                  bool               useWeights,
                                  const std::string& maxDeviation,
                                  int                symmetryRadius,
                                  bool               ignoreColinearBonds) {
  const auto moleculeVector = extractMolecules(molecules);
  const auto options        = buildOptions(useWeights, maxDeviation, symmetryRadius, ignoreColinearBonds);
  auto       gpuResult      = gpuGenerator().GetTFDMatricesGpuBuffer(moleculeVector, options);

  nvMolKit::ScopedNvtxRange range("GPU: C++ to Python tuple", nvMolKit::NvtxColor::kYellow);
  nb::list                  outputStarts;
  for (const auto outputStart : gpuResult.tfdOutputStarts) {
    outputStarts.append(outputStart);
  }

  const std::size_t totalSize = gpuResult.tfdValues.size();
  auto              array = nvMolKit::nanobind_bindings::makePyArray(gpuResult.tfdValues, nb::make_tuple(totalSize));
  return nb::make_tuple(nb::cast(std::move(array)), std::move(outputStarts));
}

}  // namespace

NB_MODULE(_TFD, module) {
  namespace nb = nanobind;
  using namespace nb::literals;

  nb::module_::import_("rdkit.Chem.rdchem");
  nb::module_::import_("nvmolkit._arrayHelpers");

  module.def("GetTFDMatricesGpuBuffer",
             &getTFDMatricesGpuBuffer,
             "mols"_a,
             "useWeights"_a          = true,
             "maxDev"_a              = "equal",
             "symmRadius"_a          = 2,
             "ignoreColinearBonds"_a = true);
}
